services/app-api/src/postgres/contractAndRates/updateDraftContractWithRates.ts

```diff
@@ -0,0 +1,375 @@
+import { findContractWithHistory } from './findContractWithHistory'
+import { NotFoundError } from '../storeError'
+import type { ContractType } from '../../domain-models/contractAndRates'
+import type { PrismaClient } from '@prisma/client'
+import type {
+    ContractFormDataType,
+    RateFormDataType,
+    RateRevisionType,
+} from '../../domain-models/contractAndRates'
+import type { StateCodeType } from 'app-web/src/common-code/healthPlanFormDataType'
+import { includeDraftRates } from './prismaDraftContractHelpers'
+import { rateRevisionToDomainModel } from './prismaSharedContractRateHelpers'
+import type { RateFormEditable } from './updateDraftRate'
+import { isEqualData } from '../../resolvers/healthPlanPackage/contractAndRates/resolverHelpers'
+
+type ContractFormEditable = Partial<ContractFormDataType>
+
+type UpdateContractArgsType = {
+    contractID: string
+    formData: ContractFormEditable
+    rateFormDatas?: RateFormDataType[]
+}
+
+const sortRatesForUpdate = (
+    ratesFromDB: RateRevisionType[],
+    ratesFormClient: RateFormDataType[]
+): {
+    upsertRates: RateFormEditable[]
+    disconnectRates: string[]
+} => {
+    const upsertRates = []
+    const disconnectRates = []
+
+    // Find rates to create or update
+    for (const rateData of ratesFormClient) {
+        // Find a matching rate revision id in the draftRatesFromDB array.
+        const matchingDBRate = ratesFromDB.find(
+            (dbRate) => dbRate.id === rateData.id
+        )
+
+        // If there are no matching rates we push into createRates
+        if (!matchingDBRate) {
+            upsertRates.push({
+                id: rateData.id,
+                ...rateData,
+            })
+            continue
+        }
+
+        // If a match is found then we deep compare to figure out if we need to update.
+        const isRateDataEqual = isEqualData(matchingDBRate.formData, rateData)
+
+        // If rates are not equal we then make the update
+        if (!isRateDataEqual) {
+            upsertRates.push({
+                id: rateData.id,
+                rateID: matchingDBRate.id,
+                ...rateData,
+            })
+        }
+    }
+
+    // Find rates to disconnect
+    for (const dbRate of ratesFromDB) {
+        //Find a matching rate revision id in the ratesFormClient
+        const matchingHPPRate = ratesFormClient.find(
+            (convertedRate) => convertedRate.id === dbRate.id
+        )
+
+        // If convertedRateData does not contain the rate revision id from DB, we push these revisions rateID in disconnectRates
+        if (!matchingHPPRate && dbRate.formData.rateID) {
```

Comment:
This is checking for the `rateID` because in the `RateFormDataType`, everything is optional. We are also using this type when converting HPP rates that do not have `rateID`s. We could add it, but that's more changes to the protobuf which I was trying to avoid.